File: compiler/testData/loadJava/compiledJava/sam/adapters/TypeParameterOfMethod.java
package test;

import java.lang.UnsupportedOperationException;
import java.util.*;
import java.util.Comparator;

public class TypeParameterOfMethod {
    public static <T> T max(Comparator<T> comparator, T konstue1, T konstue2) {
        return comparator.compare(konstue1, konstue2) > 0 ? konstue1 : konstue2;
    }

    public static <T extends CharSequence> T max2(Comparator<T> comparator, T konstue1, T konstue2) {
        return comparator.compare(konstue1, konstue2) > 0 ? konstue1 : konstue2;
    }

    public static <A extends CharSequence, B extends List<A>> void method(Comparator<A> a, B b) {
        throw new UnsupportedOperationException();
    }
}

File: compiler/testData/codegen/box/sam/adapters/typeParameterOfMethod.kt
// TARGET_BACKEND: JVM
// MODULE: lib
// FILE: WeirdComparator.java

import java.util.*;

class WeirdComparator {
    public static <T> T max(Comparator<T> comparator, T konstue1, T konstue2) {
        return comparator.compare(konstue1, konstue2) > 0 ? konstue1 : konstue2;
    }

    public static <T extends CharSequence> T max2(Comparator<T> comparator, T konstue1, T konstue2) {
        return comparator.compare(konstue1, konstue2) > 0 ? konstue1 : konstue2;
    }
}

// MODULE: main(lib)
// FILE: 1.kt

fun box(): String {
    konst result = WeirdComparator.max<String>({ a, b -> a.length - b.length }, "java", "kotlin")
    if (result != "kotlin") return "Wrong: $result"

    konst result2 = WeirdComparator.max2<String>({ a, b -> a.length - b.length }, "java", "kotlin")
    if (result2 != "kotlin") return "Wrong: $result"

    return "OK"
}
